probe_scraper/parsers/third_party/parse_events.py

```diff
@@ -0,0 +1,304 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import re
+import yaml
+import itertools
+import datetime
+import string
+from shared_telemetry_utils import add_expiration_postfix
+
+MAX_CATEGORY_NAME_LENGTH = 30
+MAX_METHOD_NAME_LENGTH = 20
+MAX_OBJECT_NAME_LENGTH = 20
+MAX_EXTRA_KEYS_COUNT = 10
+MAX_EXTRA_KEY_NAME_LENGTH = 15
+
+IDENTIFIER_PATTERN = r'^[a-zA-Z][a-zA-Z0-9_.]+[a-zA-Z0-9]$'
+DATE_PATTERN = r'^[0-9]{4}-[0-9]{2}-[0-9]{2}$'
+
+def nice_type_name(t):
+    if isinstance(t, basestring):
+        return "string"
+    return t.__name__
+
+def convert_to_cpp_identifier(s, sep):
+    return string.capwords(s, sep).replace(sep, "")
+
+class OneOf:
+    """This is a placeholder type for the TypeChecker below.
+    It signals that the checked value should match one of the following arguments
+    passed to the TypeChecker constructor.
+    """
+    pass
+
+class TypeChecker:
```

Comment:
The implementation of this class seems to be a bit odd. An alternative could be to have a set of TypeChecker classes which know how to run their own validation. Something like
```python
class AtomicTypeChecker:
    """Validate a simple value against a given type"""
    def __init__(self, instance_type):
        self.instance_type = instance_type
   
    def check(self, identifier, key, value):
        pass


class ListTypeChecker:
    """Validate a list of values against a given type"""
    def __init__(self, instance_type):
        self.instance_type = instance_type
   
    def check(self, identifier, key, value):
        pass


class DictTypeChecker:
    """Validate the keys and the values of a dict against a given type"""
    def __init__(self, keys_instance_type, values_instance_type):
        self.keys_instance_type  = keys_instance_type
        self.values_instance_type = values_instance_type
   
    def check(self, identifier, key, value):
        pass


class MultiTypeChecker:
    """Validate a simple value against a list of possible types"""
    def __init__(self, *instance_types):
        if not instance_type:
            raise ValueError("At least one instance type is required")
        self.instance_types = instance_types
        self.values_instance_type = values_instance_type
   
    def check(self, identifier, key, value):
        pass
```
This way you could still call .check on the instances of these classes, ignoring what's the actual TypeChecker class that you are using. Also, this design lets you define even more specific checkers (or validators) e.g. 
```python
class StringTypeChecker(AtomicTypeChecker):
    def __init__(min_length, max_length):
        super(StringTypeChecker, self).__init__(basestring)
        self.min_length = min_length
        self.max_length = max_length

    def check(self, identifier, key, value):
        assert isinstance(value, basestring) and self.min_length <= len(value) <= self.max_length
```